src/librustc_resolve/diagnostics.rs

```diff
@@ -444,6 +444,52 @@ impl SomeTrait for Foo { // ok!
 ```
 "##,
 
+E0406: r##"
+A function is referring to an associated type which hasn't been declared in
+the trait. Example of erroneous code:
+
+```compile_fail
+trait Foo {
+    type Bar;
+
+    // error: function signature contains a reference to `Self::Baz`, but
+    //        `Baz` hasn't been declared as an associated type of the trait
+    fn return_bool(&self, &Self::Bar, &Self::Baz) -> bool;
+}
```

Comment:
I tried searching the codebase, hoping to find answers, but like you, my knowledge of compilers isn't exactly top-notch. I suspect the compiler is showing E0220 until this lands. I asked on the IRC regrading this as well,  and it seems that E0406 *should* be thrown here.